### PR TITLE
fix: sending mail and adjusting settings FP with newer NC mail versions

### PR DESCRIPTION
**Pull request by EsadCetiner**
Some recent changes to Nextcloud Mail in the past few months have introduced a few additional false positives, this PR resolves those new false positives.

### CODE REVIEW

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 1]
```diff
@@ -1609,14 +1611,19 @@ SecRule REQUEST_FILENAME "@endsWith /apps/mail/api/contactIntegration/new" \
     setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT'"
 
 # Changing Nextcloud Mail Settings
-SecRule REQUEST_FILENAME "@rx /apps/mail/api/preferences/(?:tag-classified-messages|collect-data|account-settings|start-mailbox-id)$" \
+# Sorting email from oldest/newest
+# Adjust whether to place a reply at the top or bottom
+# Enabling/disabling external avatars from Gravtar
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1616]:
```suggestion
# Enabling/disabling external avatars from Gravatar
```

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 2]
```diff
@@ -1609,14 +1611,19 @@ SecRule REQUEST_FILENAME "@endsWith /apps/mail/api/contactIntegration/new" \
     setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT'"
 
 # Changing Nextcloud Mail Settings
-SecRule REQUEST_FILENAME "@rx /apps/mail/api/preferences/(?:tag-classified-messages|collect-data|account-settings|start-mailbox-id)$" \
+# Sorting email from oldest/newest
+# Adjust whether to place a reply at the top or bottom
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1615]:
```suggestion
# Adjusting placement of reply (top / bottom)
```

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 3]
```diff
@@ -1721,3 +1728,35 @@ SecRule REQUEST_FILENAME "@rx /apps/mail/api/mailboxes/[0-9]+$" \
     nolog,\
     ver:'nextcloud-rule-exclusions-plugin/1.0.0',\
     setvar:'tx.allowed_methods=%{tx.allowed_methods} DELETE'"
+
+# Writing an draft email
+# Email subject/body could be anything
+# PUT - Composing an draft email
+# DELETE - Discarding a draft email
+SecRule REQUEST_FILENAME "@rx /apps/mail/api/drafts(?:/[0-9]+|/\{id\})?$" \
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1736]: With `id`, isn't there any identifier? The URL would look like this: `/apps/mail/api/drafts/{id}`.

And just to be sure, you want disable those targets for:
- `.../drafts`
- `.../drafts/12345`
- `.../drafts/{id}`

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 4]
```diff
@@ -0,0 +1,167 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
+  enabled: true
+  name: 9508991.yaml
+tests:
+  - test_title: 9508991-1
+    desc: Drafing an email with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:9]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: yes
      Formatting option: disabled
      Target: json.body
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/1
+            data: "json.body=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-2
+    desc: Drafing an email with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:25]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: yes
      Formatting option: disabled
      Target: json.data.body.value
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/1
+            data: "json.data.body.value=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-3
+    desc: Drafing an email with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:41]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: yes
      Formatting option: enabled
      Target: json.data.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/1
+            data: "json.data.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-4
+    desc: Drafing an email with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:57]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: yes
      Formatting option: enabled
      Target: json.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/1
+            data: "json.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-5
+    desc: Drafing an email with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:73]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: no
      Formatting option: disabled
      Target: json.body
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.body=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-6
+    desc: Drafing an email with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:89]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: no
      Formatting option: disabled
      Target: json.data.body.value
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.data.body.value=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-7
+    desc: Drafing an email with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:105]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: no
      Formatting option: enabled
      Target: json.data.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.data.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-8
+    desc: Drafing an email with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:121]:
```suggestion
    desc: |
      Drafting an email, content could be anything.
      Draft ID set: no
      Formatting option: enabled
      Target: json.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-9
+    desc: Drafing an email, email subject could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:137]:
```suggestion
    desc: |
      Drafting an email, subject could be anything. 
      Draft ID set: no
      Target: json.subject
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.subject=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-10
+    desc: Drafing an email, email subject could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:153]:
```suggestion
    desc: |
      Drafting an email, subject could be anything.
      Draft ID set: yes
      Target: json.subject
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 14]
```diff
@@ -0,0 +1,103 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
+  enabled: true
+  name: 9508992.yaml
+tests:
+  - test_title: 9508992-1
+    desc: Sending an email with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:9]:
```suggestion
    desc: |
      Sending an email, content could be anything.
      Formatting option: enabled
      Target: json.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/outbox/1
+            data: "json.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508992-2
+    desc: Sending an email with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:25]:
```suggestion
    desc: |
      Sending an email, content could be anything.
      Formatting option: disabled
      Target: json.body
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/outbox/1
+            data: "json.body=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508992-3
+    desc: Sending an email, email subject could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:41]:
```suggestion
    desc: |
      Sending an email, subject could be anything.
      Target: json.subject
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/outbox/1
+            data: "json.subject=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508992-4
+    desc: Modifying email from outbox with "formatting" option enabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:57]:
```suggestion
    desc: |
      Modifying email from outbox, content could be anything.
      Formatting option: enabled
      Target: json.editorBody
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/outbox/from-draft/1
+            data: "json.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508992-5
+    desc: Modifying email from outbox with "formatting" option disabled, email content could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:73]:
```suggestion
    desc: |
      Modifying email from outbox, content could be anything.
      Formatting option: disabled
      Target: json.body
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/outbox/from-draft/1
+            data: "json.body=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508992-6
+    desc: Modifying email from outbox, email subject could be anything
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508992.yaml:89]:
```suggestion
    desc: |
      Modifying email from outbox, subject could be anything.
      Target: json.subject
```

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 20]
```diff
@@ -1578,14 +1578,16 @@ SecRule REQUEST_FILENAME "@rx /apps/mail/api/outbox(?:/[0-9]+)?$" \
...
+# Assigning/removing flags for emails
+# Marking an email as junk/not-junk
 SecRule REQUEST_FILENAME "@rx /apps/mail/api/messages/[0-9]+/flags$" \
     "id:9508978,\
     phase:1,\
     pass,\
     t:none,\
     nolog,\
     ver:'nextcloud-rule-exclusions-plugin/1.0.0',\
+    ctl:ruleRemoveTargetById=942290;ARGS_NAMES:json.flags.$notjunk,\
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1590]: We'll have to make an update to the linter in order to allow for the `$.

**Review by theseion — CHANGES_REQUESTED** [review]

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 20 above)
Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:1590]: https://github.com/coreruleset/secrules_parsing/pull/66

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 3 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:1736]: Yes, it's meant to match all 3. For whatever reason, sometimes the request will be made to the ``/apps/mail/api/drafts/{id}`` endpoint.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/drafts/{id}
+            data: "json.editorBody=<script>"
+          output:
+            no_log_contains: id "941101"
+  - test_title: 9508991-9
+    desc: |
+      Drafting an email, subject could be anything.
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508991.yaml:170]:
```suggestion
      Drafting an email, subject could be anything.
```

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 20 above)
Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:1590]: The PR was merged.

**Review by theseion — APPROVED** [review]

**Pull request merged**